src/metabase/util/malli/schema.clj

```diff
@@ -28,186 +29,191 @@
   Something that adheres to this schema is guaranteed to to work with `Integer/parseInt`."
-  (mc/schema
+  (mu/with-api-error-message
     [:and
      :string
-     [:fn {:error/fn (fn [_ _] (deferred-tru "value must be a valid integer."))}
-      #(u/ignore-exceptions (Integer/parseInt %))]]))
+     [:fn #(u/ignore-exceptions (Integer/parseInt %))]]
+    (deferred-tru "value must be a valid integer.")))
 
 (def IntStringGreaterThanZero
   "Schema for a string that can be parsed as an integer, and is greater than zero.
   Something that adheres to this schema is guaranteed to to work with `Integer/parseInt`."
-  (mc/schema
+  (mu/with-api-error-message
     [:and
      :string
-     [:fn {:error/fn (fn [_ _] (deferred-tru "value must be a valid integer greater than zero."))}
-      #(u/ignore-exceptions (< 0 (Integer/parseInt %)))]]))
+     [:fn #(u/ignore-exceptions (< 0 (Integer/parseInt %)))]]
+    (deferred-tru "value must be a valid integer greater than zero.")))
 
 (def IntStringGreaterThanOrEqualToZero
   "Schema for a string that can be parsed as an integer, and is greater than or equal to zero.
   Something that adheres to this schema is guaranteed to to work with `Integer/parseInt`."
-  (mc/schema
+  (mu/with-api-error-message
     [:and
      :string
-     [:fn {:error/fn (fn [_ _] (deferred-tru "value must be a valid integer greater than or equal to zero."))}
-      #(u/ignore-exceptions (<= 0 (Integer/parseInt %)))]]))
+     [:fn #(u/ignore-exceptions (<= 0 (Integer/parseInt %)))]]
+    (deferred-tru "value must be a valid integer greater than or equal to zero.")))
 
 (def BooleanString
   "Schema for a string that is a valid representation of a boolean (either `true` or `false`).
   Something that adheres to this schema is guaranteed to to work with `Boolean/parseBoolean`."
-  (mc/schema
-    [:enum "true" "false"]))
+  (mu/with-api-error-message
+    [:enum "true" "false"]
+    (deferred-tru "value must be a valid boolean string (''true'' or ''false'').")))
 
 (def TemporalString
   "Schema for a string that can be parsed by date2/parse."
-  (mc/schema
+  (mu/with-api-error-message
     [:and
      :string
-     [:fn {:error/fn (fn [_ _] (deferred-tru "value must be a valid date string"))}
-      #(u/ignore-exceptions (boolean (u.date/parse %)))]]))
+     [:fn #(u/ignore-exceptions (boolean (u.date/parse %)))]]
+    (deferred-tru "value must be a valid date string")))
 
 (def JSONString
   "Schema for a string that is valid serialized JSON."
-  (mc/schema
+  (mu/with-api-error-message
     [:and
      :string
-     [:fn {:error/fn (fn [_ _] (deferred-tru "value must be a valid JSON string."))}
-      #(try
-         (json/parse-string %)
-         true
-         (catch Throwable _
-           false))]]))
+     [:fn #(try
+             (json/parse-string %)
+             true
+             (catch Throwable _
+               false))]]
+    (deferred-tru "value must be a valid JSON string.")))
```

Comment:
I havn't looked into how, or if its possible, but do you think we can remove the “value must be” prefix everywhere, and tack it onto the top level? e.g. for JSONString:

```
(umd/describe [:map [:x JSONString] [:y JSONString]])
;; => "map where {:x -> <value must be a valid JSON string>, :y -> <value must be a valid JSON string>}"

;; seems better:
;; => "value must be a map where {:x -> <valid JSON string>, :y -> <valid JSON string>}"
```

Reply:
Trying it out here : https://github.com/metabase/metabase/pull/28208

WDYT?